### PR TITLE
Tywarren/sc 3278/connection monitoring via ping

### PR DESCRIPTION
**Pull request by jackofsometrade**

### CODE REVIEW

File: src/API.ts [hunk 1]
```diff
@@ -13,11 +13,11 @@ export enum ParticipantEventDetail {
   ConnectionError = "connection_error",
 }
 
-export type Participant = {
+export interface Participant {
   type: "inmate" | "doc" | "user";
   id: string;
-  detail: ParticipantEventDetail | undefined;
-};
+  detail?: ParticipantEventDetail;
```
Comment (jackofsometrade) [inline src/API.ts:19]: switching to interface and using optional `?` allows consuming clients to ignore optional `detail`

**Comment by jackofsometrade** [comment]
Notable things:  

- Hard-coded interval of 500ms (keep in mind this is not HTTP, but websocket)
- Hard-coded TTL of 5 seconds
- Possibility of noisiness if client connection average is near a border between 2 neighboring quality ranges

File: src/RoomClient.ts [hunk 2]
```diff
@@ -216,6 +231,14 @@ class RoomClient {
     this.emitter.off(name, handler);
   }
 
+  get ConnectionState(): ConnectionState {
```
Comment (cainlevy) [inline src/RoomClient.ts:234]: camelCase convention for a property:

```suggestion
  get connectionState(): ConnectionState {
```

File: src/RoomClient.ts [hunk 3]
```diff
@@ -265,6 +288,8 @@ class RoomClient {
     this.producers.audio?.close();
     this.producers.video?.close();
     this.emitter.all.clear();
+    this.client.connectionMonitor.emitter.all.clear();
```
Comment (cainlevy) [inline src/RoomClient.ts:291]: can this be handled by the `stop()` as a private detail?

File: src/ConnectionMonitor.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+      this._currentQuality = q;
+    }
+  }
+
+  /**
+   * starts the monitor
+   */
+  start() {
+    if (!this.timer) {
+      this.socket.on(PONG_EVENT, this.handleResponse);
```
Comment (cainlevy) [inline src/ConnectionMonitor.ts:120]: does CVH need a handler for this? did i miss a PR, or is this some feature of socket.io?

File: src/useConnectCall.ts [hunk 5]
```diff
@@ -99,12 +105,24 @@ const useConnectCall = ({
...
+      msElapsed: number;
+    }) => {
       onTimer && onTimer(name, msRemaining, msElapsed);
     },
     [onTimer]
   );
 
+  const handleConnectionState = (connectionState: ConnectionState) => {
+    onConnectionState && onConnectionState(connectionState);
+  };
```
Comment (cainlevy) [inline src/useConnectCall.ts:124]: we should apply `useCallback` here to memoize the function so it's suitable for dependency arrays in the caller

**Review by cainlevy — CHANGES_REQUESTED** [review]

File: src/API.ts (hunk 1 above)
Comment (jackofsometrade) [reply src/API.ts:19]: Seems to with our current TS settings.  I didn't dig too deep.

File: src/RoomClient.ts (hunk 3 above)
Comment (jackofsometrade) [reply src/RoomClient.ts:291]: Yes!  Thanks for that, this was older code that I forgot to update.

File: src/ConnectionMonitor.ts (hunk 4 above)
Comment (jackofsometrade) [reply src/ConnectionMonitor.ts:120]: Yes, there's a PR coming shortly.

File: src/useConnectCall.ts (hunk 5 above)
Comment (jackofsometrade) [reply src/useConnectCall.ts:124]: Will do

File: src/ConnectionMonitor.ts (hunk 4 above)
Comment (jackofsometrade) [reply src/ConnectionMonitor.ts:120]: https://github.com/Ameelio/connect-video-handler/pull/200

File: src/useConnectCall.ts (hunk 5 above)
Comment (jackofsometrade) [reply src/useConnectCall.ts]: onConnectionState should only be called when there's a change in values anyway, but I like that I learned something new today.

**Review by cainlevy — APPROVED** [review]

**Pull request merged**